src/Compilers/CSharp/Portable/Compiler/Instrumentation.cs

```diff
@@ -0,0 +1,350 @@
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.CodeAnalysis;
+using Microsoft.CodeAnalysis.CodeGen;
+using Microsoft.CodeAnalysis.CSharp.Symbols;
+
+namespace Microsoft.CodeAnalysis.CSharp
+{
+    internal sealed class Instrumentation
+    {
+        private static MethodSymbol _createPayload = null;
+        private static MethodSymbol _flushPayload = null;
+
+        internal static BoundBlock InjectInstrumentation(MethodSymbol method, BoundBlock methodBody, int methodOrdinal, TypeCompilationState compilationState, CSharpCompilation compilation, DiagnosticBag diagnostics, DebugDocumentProvider debugDocumentProvider, out ImmutableArray<SourceSpan> dynamicAnalysisSpans)
+        {
+            if (methodBody != null && (object)method != GetCreatePayload(compilation) && (object)method != GetFlushPayload(compilation))
+            {
+                MethodSymbol createPayload = GetCreatePayload(compilation);
+                MethodSymbol flushPayload = GetFlushPayload(compilation);
+                if (createPayload != null && flushPayload != null)
+                {
+                    // Create the symbol for the instrumentation payload.
+                    SyntheticBoundNodeFactory factory = new SyntheticBoundNodeFactory(method, methodBody.Syntax, compilationState, diagnostics);
+                    TypeSymbol boolType = compilation.GetSpecialType(SpecialType.System_Boolean);
+                    TypeSymbol payloadElementType = boolType;
+                    ArrayTypeSymbol payloadType = ArrayTypeSymbol.CreateCSharpArray(compilation.Assembly, payloadElementType);
+                    FieldSymbol payloadField = GetPayloadField(method, methodOrdinal, payloadType, factory, compilation);
+
+                    // Synthesize the instrumentation and collect the spans of interest.
+
+                    ArrayBuilder<SourceSpan> spansBuilder = ArrayBuilder<SourceSpan>.GetInstance();
+                    BoundTreeRewriter collector = new InstrumentationInjectionWalker(method, spansBuilder, payloadField, compilationState, diagnostics, debugDocumentProvider);
+                    BoundBlock newMethodBody = (BoundBlock)collector.Visit(methodBody);
+                    dynamicAnalysisSpans = spansBuilder.ToImmutableAndFree();
+
+                    // Synthesize the initialization of the instrumentation payload array. It should be done either statically or with concurrency-safe code:
+                    //
+                    // if (payloadField == null)
+                    //     Instrumentation.CreatePayload(mvid, method, ref payloadField, payloadLength);
+
+                    // ToDo: The containing module's mvid should be computed statically and stored in a static field rather than being
+                    // recomputed in each method prologue.
+                    BoundExpression mvid = factory.Property(factory.Property(factory.TypeofBeforeRewriting(method.ContainingType), "Module"), "ModuleVersionId");
```

Comment:
Consider filing a catch-all bug like "Improve instrumentation perf" and including the URL in the TODO.